### DISCUSSION TITLE
Which is the fastest way to transfer the pdf pages into separate jpeg images?

### DISCUSSION THREAD

**Discussion by YushuoGuan** [pymupdf/PyMuPDF#1648] [Q&A] [answered]
Dear developers, 

      I use PyMuPDF to transfer pdf pages into separate images (`1 page -> 1 image`), and I hope to know if I wrote the appropriate code. 
      I have two requirements for the images: 1. direct output image bytes. 2. the format is `JPEG` instead of `PNG` (I heard opencv reading jpeg is faster that png.)
      The code are something like:

```python

doc = fitz.open('xxx.pdf') 
p0 = doc.load_page(0)
p0.save('page_0.jpg')
with open('page_0.jpg', 'rb') as infile:
	img_bytes = infile.read()     

```
       I know I could use `p0.tobytes()` to get a PNG image bytes, but I don't know how to directly generate `JPEG image bytes.`

**Comment by JorjMcKie** [comment]
Welcome and let me first transform this issue to a Discussions item.

**Comment by JorjMcKie** [comment, marked as answer]
To make an image of a page, you must first take "picture", which is called "pixmap": `pix = page.get_pixmap()`. You can influence many properties of that pixmap, like resolution, whether gray or RGB, rotation or flipping.
Once you have that, either use
* `pix.save(...)` which can only produce PNG, PNM and a handful of other formats, or
* `pix.pil_save()` which uses pillow with an additional internal step. This allows JPEG output among others.

With `.pil_save()` you have the full abilities of pillow output available: the method arguments are those of pillow, which you therefore must lookup. Simple example:
```python
page = doc[0]
pix = page.get_pixmap(dpi=300)  # RGB with a high resolution
pix.pil_save("page-0.jpg")
```

**Reply by JorjMcKie** [reply]
Please also note that the above is not restricted to PDF files: it works the same way for all supported document types like XPS, EPUB, FB2, CBZ, ...

**Comment by YushuoGuan** [comment]
Thanks for the quick reply. Is there a way to get the image bytes directly? In your solution, I have to `1. save to the disk; 2. read the bytes from the disk`, but I think the IO operation is abundant.

**Reply by JorjMcKie** [reply]
Of course there is 😁!
Instead of `pix.save` use `pix.tobytes`, and instead of `pix.pil_save` use `pix.pil_tobytes`.
Both cases produce a bytes object of the image.

**Reply by YushuoGuan** [reply]
Got it, thanks for your help~
